### DISCUSSION TITLE
NCAR-NCL installation issue with a MacPro M1

### DISCUSSION THREAD

**Discussion by lese66** [Homebrew/discussions#3573] [Casks]
Hi,

I installed NCL (NCAR software) using this formulae:

```
brew install --cask ncar-ncl

```

It ran with no issues. My brew was/is updated. However, when I type ncl in the terminal I get the following error:

```
dyld[67721]: Library not loaded: '/usr/local/lib/libgfortran.4.dylib'
  Referenced from: '/opt/homebrew/ncl-6.6.2/bin/ncl'
  Reason: tried: '/usr/local/lib/libgfortran.4.dylib' (no such file), './libgfortran.4.dylib' (no such file), './libgfortran.4.dylib' (no such file), './libgfortran.4.dylib' (no such file), '/libgfortran.4.dylib' (no such file)
Abort trap: 6
```

I am using a MacPro M1 chip. OSX 12.5.1 updated
The command `env  | grep NCARG` gives the following output: `NCARG_ROOT=/opt/homebrew/ncl-6.6.2`

Could someone be so kind to give me a hint why is this happeing? Why the libgfortran.4.dylib is not available in my system?

which gfortran
/opt/homebrew/bin/gfortran

gfortran -v

Using built-in specs.
```
COLLECT_GCC=gfortran
COLLECT_LTO_WRAPPER=/opt/homebrew/Cellar/gcc/12.1.0/bin/../libexec/gcc/aarch64-apple-darwin21/12/lto-wrapper
Target: aarch64-apple-darwin21
Configured with: ../configure --prefix=/opt/homebrew/opt/gcc --libdir=/opt/homebrew/opt/gcc/lib/gcc/current --disable-nls --enable-checking=release --with-gcc-major-version-only --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-12 --with-gmp=/opt/homebrew/opt/gmp --with-mpfr=/opt/homebrew/opt/mpfr --with-mpc=/opt/homebrew/opt/libmpc --with-isl=/opt/homebrew/opt/isl --with-zstd=/opt/homebrew/opt/zstd --with-pkgversion='Homebrew GCC 12.1.0' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --with-system-zlib --build=aarch64-apple-darwin21 --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.1.0 (Homebrew GCC 12.1.0) 
```

Thanks for any hint.

Luis

**Comment by prabshr** [comment]
I ran into similar issue with a MacPro M2
I also set the following in .zprofile
export DYLD_FALLBACK_LIBRARY_PATH="/opt/homebrew/Cellar/gcc/12.2.0/lib/gcc/current"

However, the folder has libgfortran.5.dylib rather than libgfortran.4.dylib
/opt/homebrew/Cellar/gcc/12.2.0/lib/gcc/current/libgfortran.5.dylib

So, maybe need to install older version of gcc to get the ncl working

**Comment by zmlabe** [comment]
Any updates on these issues? I have tried installing NCL with brew, macports, and conda, but all with different error messages (with Macbook Pro, M2 Pro). Thank you!

**Comment by SMillerDev** [comment]
This is an upstream bug where the precompiled binaries upstream provides depend on an x86 GFortran install. This should be reported and fixed upstream, the comment about multiple install methods having this issue is a very clear indication that this isn't a new issue.

**Comment by emresalkim** [comment]
Hi to all, are there any updates on the issue? I tried installing with miniconda. I think the errors are persisting.

**Reply by zmlabe** [reply]
Unfortunately, I still haven't had any success either.

**Reply by derekjwilliams** [reply]
Are the maintainers aware of this issue?   I've successfully installed ncdump using brew, e.g. `/opt/homebrew/bin/ncdump global_timeseries_measures.nc`works after install of ncdump, but having no luck with ncl


Install steps for ncdump:

```
brew tap homebrew/science

brew install netcdf

brew install moffat/sciencebits/cdo

brew untap moffat/sciencebits

brew install nco
```

The above steps were not without issue, e.g. there were complaints about deprecated casks "Error: homebrew/science was deprecated. This tap is now empty and all its contents were either deleted or migrated."

But ncdump did work, which was a bit surprising.  

## Attempt at installing ncl

brew install ncar-ncl:

```==> Downloading https://formulae.brew.sh/api/cask.jws.json
###################################################################################### 100.0%
==> Caveats
To use ncar-ncl, you must add the ${NCARG_ROOT}/bin directory
to your PATH environment variable.

For bash shell, add these lines to ~/.bash_profile:

  export NCARG_ROOT="/opt/homebrew/ncl-6.6.2"
  export PATH="${NCARG_ROOT}/bin:${PATH}"

You may also need to modify your DYLD_FALLBACK_LIBRARY_PATH
environment variable:

export DYLD_FALLBACK_LIBRARY_PATH=$(dirname $(gfortran --print-file-name libgfortran.3.dylib)):$DYLD_FALLBACK_LIBRARY_PATH

For other information, please see:
https://www.ncl.ucar.edu/Download/macosx.shtml

==> Downloading https://raw.githubusercontent.com/Homebrew/homebrew-cask/6de9b2827502ac665ce1
###################################################################################### 100.0%
==> Downloading https://www.earthsystemgrid.org/api/v1/dataset/ncl.662.dap/file/ncl_ncarg-6.6
###################################################################################### 100.0%
==> Downloading https://formulae.brew.sh/api/formula.jws.json
###################################################################################### 100.0%
==> Installing dependencies: xquartz
==> Downloading https://github.com/XQuartz/XQuartz/releases/download/XQuartz-2.8.5/XQuartz-2.
==> Downloading from https://objects.githubusercontent.com/github-production-release-asset-2e
###################################################################################### 100.0%
==> Installing Cask xquartz
==> Running installer for xquartz with sudo; the password may be necessary.
Password:
installer: Package name is 
installer: Installing at base path /
installer: The install was successful.
installer: The install requires logging out now.
🍺  xquartz was successfully installed!
==> Installing Cask ncar-ncl
==> Moving Generic Artifact 'include' to '/opt/homebrew/ncl-6.6.2/include'
==> Moving Generic Artifact 'bin' to '/opt/homebrew/ncl-6.6.2/bin'
==> Moving Generic Artifact 'lib' to '/opt/homebrew/ncl-6.6.2/lib'
🍺  ncar-ncl was successfully installed!
```

However when attempting to run get

```
dyld[22381]: Library not loaded: /usr/local/lib/libgfortran.4.dylib
  Referenced from: <6E838C70-A577-3DA9-AFA9-EF75ED781A06> /opt/homebrew/ncl-6.6.2/bin/ncl
  Reason: tried: '/usr/local/lib/libgfortran.4.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/local/lib/libgfortran.4.dylib' (no such file), '/usr/local/lib/libgfortran.4.dylib' (no such file), '/usr/lib/libgfortran.4.dylib' (no such file, not in dyld cache)
zsh: abort      ncl
```

Created a link: sudo ln /opt/homebrew/Cellar/gcc/13.2.0/lib/gcc/current/libgfortran.5.dylib /usr/local/lib/libgfortran.4.dylib


### Wrong architecture 

Then received a complaint about architecture, 

```
Library not loaded: /usr/local/lib/libgfortran.4.dylib
  Referenced from: <6E838C70-A577-3DA9-AFA9-EF75ED781A06> /opt/homebrew/ncl-6.6.2/bin/ncl
  Reason: tried: '/usr/local/lib/libgfortran.4.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')
```

So it looks like the brew install of ncar-ncl installs the x86_64 version for some reason.

**Comment by luoxiao11** [comment]
Manually download an early version of gcc from https://hpc.sourceforge.net/, for example, gcc-7.3. 
Unzip the downloaded file and update the ".bashrc" file with the following line: "export DYLD_FALLBACK_LIBRARY_PATH=/Usr/opt/gcc-7.3/local/lib" (make sure to adjust the path according to your local directory structure).

**Reply by bthrasher** [reply]
Oh my goodness, this worked. Thank you!!

**Reply by emresalkim** [reply]
Can you help me with the unzipping part? Is there a specific path I should unzip the downloaded gcc version?

**Reply by bthrasher** [reply]
As stated above, just point DYLD_FALLBACK_LIBRARY_PATH to wherever you unzip it to. I put mine in /opt/homebrew, so I have this:
export DYLD_FALLBACK_LIBRARY_PATH=/opt/homebrew/usr/local/lib

**Reply by emresalkim** [reply]
I moved the unzipped folder called "usr" to the /opt/homebrew. created a ".bashrc" under home and updated it with "export DYLD_FALLBACK_LIBRARY_PATH=/opt/homebrew/usr/local/lib". however ncl -V returns with "zsh: command not found: ncl" -- I'm sure I am doing sth wrong. Sorry for being this dumb 😅

**Reply by bthrasher** [reply]
You need to switch your shell from zsh to bash.

**Reply by emresalkim** [reply]
Thank you for your answer. Actually switching wasn't needed.

First I downloaded gcc-7.3 and unzipped it to my Documents folder. Then edited zhrc by adding,
`export DYLD_FALLBACK_LIBRARY_PATH=/Users/emre/Documents/usr/local/lib`

then of course, I sourced the zhrc with,
`source ~/.zshrc`

Voila! 

Once again thank you @luoxiao11 🙏

**Comment by gaelforget** [comment]
related to https://github.com/MITgcm/MITgcm/issues/846 ?

**Comment by gitleviglenn** [comment]
I am also having problems getting ncar-ncl to work on an M3 MacBook Pro running Sonoma (14.5).   I tried initially with anaconda, following the recommended method from the ncl page.   That does not work for me.   Either the executable is not installed, or it cannot be found.    The command 'ncl' or 'ncl -V' returns 'command not found'.     

So now I am trying to use brew to install.  

brew install --cask ncar-ncl
`==> Auto-updating Homebrew...
Adjust how often this is run with HOMEBREW_AUTO_UPDATE_SECS or disable with
HOMEBREW_NO_AUTO_UPDATE. Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
==> Downloading https://ghcr.io/v2/homebrew/portable-ruby/portable-ruby/blobs/sha256:303bed4c7fc431a685db3c3c151d873740114adbdccd23762ea2d1e39ea78f47
######################################################################### 100.0%
==> Pouring portable-ruby-3.3.6.arm64_big_sur.bottle.tar.gz
==> Auto-updated Homebrew!`
.
.
.
`==> Installing Cask ncar-ncl
==> Moving Generic Artifact 'include' to '/opt/homebrew/ncl-6.6.2/include'
==> Moving Generic Artifact 'bin' to '/opt/homebrew/ncl-6.6.2/bin'
==> Moving Generic Artifact 'lib' to '/opt/homebrew/ncl-6.6.2/lib'
🍺  ncar-ncl was successfully installed!
==> Caveats
==> ncar-ncl
To use ncar-ncl, you must add the ${NCARG_ROOT}/bin directory
to your PATH environment variable.

For bash shell, add these lines to ~/.bash_profile:

  export NCARG_ROOT="/opt/homebrew/ncl-6.6.2"
  export PATH="${NCARG_ROOT}/bin:${PATH}"

You may also need to modify your DYLD_FALLBACK_LIBRARY_PATH
environment variable:

export DYLD_FALLBACK_LIBRARY_PATH=$(dirname $(gfortran --print-file-name libgfortran.3.dylib)):$DYLD_FALLBACK_LIBRARY_PATH`

I then followed the recommendation to add the two lines of code to ~/.bash_profile.   ncl does not work.   Nothing happens upon trying to open it.   So then I modified `DYLD_FALLBACK_LIBRARY_PATH` as recommended.   Now, when trying to open/run ncl, apparently my system can find ncl but I get a pop up window stating that 

`"ncl" cannot be opened because the developer cannot be verified.  macOS cannot verify that this app is free from malware.`

What now?   I have wasted a large amount of time on this.

**Reply by gitleviglenn** [reply]
Also, to add detail to the above account, at no point of my attempts on this M3 MacBook did I run into errors related to my version of the gcc compiler.

**Comment by bthrasher** [comment]
https://nordvpn.com/blog/macos-cannot-verify-this-app-is-free-from-malware/

**Comment by gitleviglenn** [comment]
Ok, as an update.   Once I allowed my Mac to open up the ncl package from homebrew, then I ended up with the gfortran that others on this thread mentioned.   I downloaded gcc-7.3 from  https://hpc.sourceforge.net/ and untarred the package.   I was running into permission errors when trying to locate it in /usr/opt/ or /opt/homebrew/.   If I untarr in a location for which I have permission and then add an appropriate DYLD_FALLBACK_LIBRARY_PATH definition in my .bash_profile file ncl seems to be working.    However, this is only after my Mac not wanting to open numerous (~10) packages within the gcc-7.3 package and me overriding the safety recommendations.      Thanks for your help!    Especially luoxiao11!
